device/alpaka/src/seeding/spacepoint_binning.cpp

```diff
@@ -0,0 +1,164 @@
+
+// VecMem include(s).
+#include <vecmem/utils/copy.hpp>
+#include <vecmem/utils/cuda/copy.hpp>
+
+namespace traccc::alpaka {
+
+spacepoint_binning::spacepoint_binning(
+    const seedfinder_config& config, const spacepoint_grid_config& grid_config,
+    const traccc::memory_resource& mr)
+    : m_config(config),
+      m_axes(get_axes(grid_config, (mr.host ? *(mr.host) : mr.main))),
+      m_mr(mr) {
+
+    // Initialize m_copy ptr based on memory resources that were given
+    if (mr.host) {
+        m_copy = std::make_unique<vecmem::cuda::copy>();
+    } else {
+        m_copy = std::make_unique<vecmem::copy>();
+    }
+}
+
+// Grid Capacity Kernel
+struct CountGridCapacityKernel {
+    template <typename Acc>
+    ALPAKA_FN_ACC void operator()(
+        Acc const& acc,
+        const seedfinder_config& config,
+        const sp_grid::axis_p0_type& phi_axis,
+        const sp_grid::axis_p1_type& z_axis,
+        const spacepoint_collection_types::const_view& spacepoints_view,
+        vecmem::data::vector_view<unsigned int>& grid_capacities_view
+    ) const
+    {
+        auto const globalThreadIdx = ::alpaka::getIdx<::alpaka::Grid, ::alpaka::Threads>(acc)[0u];
+        device::count_grid_capacities(globalThreadIdx, config,
+                                      phi_axis, z_axis, spacepoints_view,
+                                      grid_capacities_view);
+    }
+};
+
+// Populate Grid Kernel
+struct PopulateGridKernel {
+    template <typename Acc>
+    ALPAKA_FN_ACC void operator()(
+        Acc const& acc,
+        const seedfinder_config& config,
+        const spacepoint_collection_types::const_view& spacepoints_view,
+        sp_grid_view& grid_view
+    ) const
+    {
+        auto const globalThreadIdx = ::alpaka::getIdx<::alpaka::Grid, ::alpaka::Threads>(acc)[0u];
+
+        // Check if anything needs to be done.
+        const spacepoint_collection_types::const_device spacepoints(
+            spacepoints_view);
+
+        if (globalThreadIdx >= spacepoints.size()) {
+            return;
+        }
+        const spacepoint sp = spacepoints.at(globalThreadIdx);
+
+        /// Check out if the spacepoint can be used for seeding.
+        if (is_valid_sp(config, sp) != detray::detail::invalid_value<size_t>()) {
+
+            // Set up the spacepoint grid object(s).
+            sp_grid_device grid(grid_view);
+            const sp_grid_device::axis_p0_type& phi_axis = grid.axis_p0();
+            const sp_grid_device::axis_p1_type& z_axis = grid.axis_p1();
+
+            // Find the grid bin that the spacepoint belongs to.
+            const internal_spacepoint<spacepoint> isp(sp, globalThreadIdx,
+                                                      config.beamPos);
+            const std::size_t bin_index =
+                phi_axis.bin(isp.phi()) + phi_axis.bins() * z_axis.bin(isp.z());
+
+            // Add the spacepoint to the grid.
+            grid.bin(bin_index).push_back(std::move(isp));
+        }
```

Comment:
This I need to check again....I was originally getting errors around calling a `__host__` function from  `__host__ __device__` function here I believe, so I inlined it just to test some other parts. I believe I ended up with it fixed in my full seeding example, so I expect I was just doing something else wrong here.